File: garbage/solution/garbage-sub_3_cc.py
N_LIMIT = 2000

class point:
    def __init__(self, x, y, w):
        self.x = x
        self.y = y
        self.w = w

# read input
N, W, H = map(int, input().split())
pts = []
coords_x = []
coords_y = []
for i in range(N):
    x, y, w = map(int, input().split())
    pts.append(point(x, y, w))
    coords_x.append(x)
    coords_x.append(x + W)
    coords_y.append(y)
    coords_y.append(y + H)
coords_x.sort()
coords_y.sort()

# coordinate compression
compress_x = {}
compress_y = {}
for e in coords_x: compress_x.update({e: len(compress_x)})
for e in coords_y: compress_y.update({e: len(compress_y)})
grid = [[0 for i in range(2*N_LIMIT+1)] for j in range(2*N_LIMIT+1)] # can be optimized
for i in range(N):
    grid[compress_x[pts[i].x]+1][compress_y[pts[i].y]+1] += pts[i].w

# 2D prefix sum
for i in range(1, 2*N_LIMIT+1):
    for j in range(1, 2*N_LIMIT+1):
        grid[i][j] += grid[i-1][j] + grid[i][j-1] - grid[i-1][j-1]

# check all compressed positions as lower left corner
result = 0
for x in coords_x:
    if (x+W not in compress_x): continue
    for y in coords_y:
        if (y+H not in compress_y): continue
        value = grid[compress_x[x+W]][compress_y[y+H]]
        value -= grid[compress_x[x]][compress_y[y+H]]
        value -= grid[compress_x[x+W]][compress_y[y]]
        value += grid[compress_x[x]][compress_y[y]]
        result = max(result, value)
print(result)

File: garbage/solution/garbage-sub_3_cc.cpp
#include <bits/stdc++.h>
using namespace std;

typedef long long ll;

#define N_LIMIT 2000

int main() {
    // read input
    int N, W, H;
    cin >> N >> W >> H;
    vector<pair<int, int>> pts;
    vector<int> weights;
    set<int> coords_x, coords_y;
    for (int i = 0; i < N; i++) {
        int x, y, w;
        cin >> x >> y >> w;
        pts.push_back({x, y});
        weights.push_back(w);
        coords_x.insert(x);
        coords_x.insert(x + W);
        coords_y.insert(y);
        coords_y.insert(y + H);
    }

    // coordinate compression
    map<int, int> compress_x, compress_y;
    for (auto &e : coords_x) compress_x[e] = compress_x.size();
    for (auto &e : coords_y) compress_y[e] = compress_y.size();
    ll grid[2*N_LIMIT+1][2*N_LIMIT+1];
    for (int i = 0; i < N; i++) {
        grid[compress_x[pts[i].first]+1][compress_y[pts[i].second]+1] += weights[i];
    }

    // 2D prefix sum
    for (int i = 1; i <= 2*N_LIMIT; i++) {
        for (int j = 1; j <= 2*N_LIMIT; j++) {
            grid[i][j] += grid[i-1][j] + grid[i][j-1] - grid[i-1][j-1];
        }
    }

    // check all compressed positions as lower left corner
    ll result = 0;
    for (auto &x : coords_x) {
        if (!compress_x.count(x+W)) continue;
        for (auto &y : coords_y) {
            if (!compress_y.count(y+H)) continue;
            ll value = grid[compress_x[x+W]][compress_y[y+H]]
                     - grid[compress_x[x]][compress_y[y+H]]
                     - grid[compress_x[x+W]][compress_y[y]]
                     + grid[compress_x[x]][compress_y[y]];
            result = max(result, value);
        }
    }
    cout << result << endl;
}
